sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,968 @@
+    }
+
+    /**
+     * Returns InputFormat object.
+     */
+    @VisibleForTesting
+    InputFormat<?, ?> getInputFormat(){
+      return inputFormatObj;
+    }
+
+    @Override
+    public Coder<KV<K, V>> getDefaultOutputCoder() {
+      return KvCoder.of(keyCoder, valueCoder);
+    }
+
+    @Override
+    public BoundedReader<KV<K, V>> createReader(PipelineOptions options) throws IOException {
+      this.validate();
+      if (inputSplit == null) {
+          throw new IOException(HadoopInputFormatIOContants.CREATEREADER_UNSPLIT_SOURCE_ERROR_MSG);
+      } else {
+        return new HadoopInputFormatReader<Object>(this, keyTranslationFunction,
+            valueTranslationFunction, inputFormatObj, inputSplit.getSplit());
+      }
+    }
+
+    /** BoundedReader for HadoopInputFormatSource. */
+    class HadoopInputFormatReader<T extends Object> extends BoundedSource.BoundedReader<KV<K, V>> {
+
+      private final HadoopInputFormatBoundedSource<K, V> source;
+      private final InputSplit split;
+      @Nullable private final SimpleFunction<T, K> keyTranslationFunction;
+      @Nullable private final SimpleFunction<T, V> valueTranslationFunction;
+      private volatile boolean doneReading = false;
+      private long recordsReturned = 0L;
+      private InputFormat<?, ?> inputFormatObj;
+
+      public HadoopInputFormatReader(HadoopInputFormatBoundedSource<K, V> source,
+          @Nullable SimpleFunction keyTranslationFunction,
+          @Nullable SimpleFunction valueTranslationFunction,
+          InputFormat<?, ?> inputFormatObj,
+          InputSplit split) {
+        this.source = source;
+        this.keyTranslationFunction = keyTranslationFunction;
+        this.valueTranslationFunction = valueTranslationFunction;
+        this.inputFormatObj = inputFormatObj;
+        this.split = split;
+      }
+
+      @Override
+      public HadoopInputFormatBoundedSource<K, V> getCurrentSource() {
+        return source;
+      }
+
+      private RecordReader<T, T> currentReader;
+      private KV<K, V> currentRecord;
+
+      @Override
+      public boolean start() throws IOException {
+        try {
+          TaskAttemptContextImpl attemptContext = new TaskAttemptContextImpl(
+              source.getConfiguration().getHadoopConfiguration(), new TaskAttemptID());
+          currentReader =
+              (RecordReader<T, T>) inputFormatObj.createRecordReader(split, attemptContext);
+            currentReader.initialize(split, attemptContext);
+            if (currentReader.nextKeyValue()) {
+              currentRecord = nextPair();
+              return true;
+            }
+        } catch (InterruptedException e) {
+          throw new IOException("Unable to read data : ", e);
+        }
+        currentReader = null;
+        currentRecord = null;
+        doneReading = true;
+        return false;
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        try {
+          synchronized (currentReader) {
```

Comment:
When we debugged, till advance() method, multiple threads were not generated. We saw multiple threads only for advance() and getFractionConsumed(), hence we didn't add the synchronized block on currentReader in start() method, as we didn't see multiple threads accessing currentReader at that point.